### DISCUSSION TITLE
Visualising Propulsion System

### DISCUSSION THREAD

**Discussion by AstroYak** [JSBSim-Team/jsbsim#951] [Q&A]
Hello All, 

Forgive me, i am very far from a programmer, but i've been working with JSBSim for about a year now and i'm slowly getting better. 

Recently i was working on the helicopters rotor controls, and it helped me to try and trace back the variable to see all of the different things that interract with it and i made something of a mind map or a flow chart. I have attached it for you to look at if you like. 

I'm currently trying to do the same thing with the propulsion system but i haven't had much success at all, i find myself getting stuck in loops or with circular referencing. If anybody is able to help me out, i'd greatly appreciate it. The key areas of intrest for me are as follows: 

- ITT
- Prop/Engine RPM 
- Prop/Engine Torque
- Fuel Flow

Let me know what you think and thanks for your help!

![RotorControlVisualised](https://github.com/JSBSim-Team/jsbsim/assets/139533907/7f145da1-9e50-4b93-ba8a-b666c4f36842)

**Comment by seanmcleod** [comment]
@AstroYak useful looking flow chart. Can you post a version saved as PNG rather than JPEG? The JPEG compression makes it quite difficult to read the text.

Talking of these sorts of flow charts for describing and understanding the flight control system there is a tool that was developed, called JSBSimCommander - https://github.com/JSBSim-Team/jsbsimcommander quite a while ago. However it hasn't been updated in a very long time.

In terms of having people help out in understanding the propulsion system it's probably best to write up/sketch your current understanding with a couple of initial questions.

**Reply by AstroYak** [reply]
Hey @seanmcleod, 

Thanks so much for getting back to me. Really appreciate it :) 

Regarding the PNG, absolutely, I'll attach it at the bottom of this. 
Regarding the questions, also absolutely! 

I think a large part of my confusion is due to my lack of understanding of basic software frameworks and architectures. Perhaps if I walk you through my process with an example, you can understand my thinking and point out some vulnerabilities, and maybe suggest a better alternative. 

Example:
 - I am hoping to have better control over some properties of the aircraft. Let's start with _propulsion/engine/itt-c_.

 - Rather than applying a multiplier or something to the property, I first want to understand how it is derived. I go to FGTurboProp.cpp and see that it is tied to _&Eng_ITT_degC_

 - I trace this back to double _FGTurboProp::Run(void)_ and find this: 
 _Eng_ITT_degC  = ExpSeek(&Eng_ITT_degC,ITT_goal,ITT_Delay,ITT_Delay*1.2)_

 - I understand this to mean that a delay will lead the variable towards _ITT_goal_ which I find is written here
 _double ITT_goal = ITT_N1->GetValue((N1-old_N1)*300+N1,1);_
 
  - I trace _ITT_N1_ to a table in the engine xml files that takes inputs from (N1-old_N1)*300+N1 and 1. 
  
  In my head, I can then change that table as I need to to try and get my properties to line up with experimental data. 
  I'm not sure if what I did is improper or a bit rough or anything, but I hope you can understand where I'm coming from. 
  When I try to follow a similar procedure with something like RPM or torque, I have less success.
  
  Example:
   - I'm trying to track _propulsion/engine/propeller-rpm_ and I see that _FGPropeller.cpp_ calls _GetRPM_. Which in _Propeller.h_ returns _RPM_ 
   -  I see in FGPropeller.cpp that in the constructor the default value _RPM = 0;_ however I assume that this is overwritten when FGPropeller::Calculate is called. 
   
   - I notice that it's written that "we can derive the torque on the propeller and its acceleration to give a new RPM" and i think this is done a few lines down here: _RPM = (RPS + ((ExcessTorque / Ixx) / (2.0 * M_PI)) * deltaT) * 60.0_
   
   - Then I track further and find _ExcessTorque = PowerAvailable / omega;_ and later arrive here: _ PowerAvailable = EnginePower - GetPowerRequired();_
   
   - This is where I really lose myself. I'm pretty sure that EnginePower is supposed to be given when the calculate function is called but I can't find any instance of it being called? So I can't work out what is the key parameter that is driving the change in RPM (which to me seems like it then influences everything else.)
   
Does this approach make sense to you, or is there a better way to come out with an understanding of how each and all of the properties and variables interact with each other? Let me know and I'm so sorry if my writing or thought process is confusing/primitive, I really can't stress enough how much I'm new to all this! 

Thanks for all your help and looking forward to hearing from you

![flowchart](https://github.com/JSBSim-Team/jsbsim/assets/139533907/9b20302d-87c2-4d3e-8af4-438566044172)

**Comment by seanmcleod** [comment]
I think the approach you're taking makes sense, I follow a similar approach. I'd just add two additional things that I do.

-  Look at textbooks such as Stevens and Lewis "Aircraft Control and Simulation" to get some idea on how some of these systems are modelled. Will give you a better understanding of what some of the variables refer to, e.g. when modelling a propeller etc.

- Set some breakpoints and step through the code. You might miss some things if you only read the code.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/b93f5cd0-c813-4425-87e3-cc22befd34b1)

> I'm pretty sure that EnginePower is supposed to be given when the calculate function is called but I can't find any instance of it being called? 

So you can see in the screen snippet that `FGPiston`, modelling a piston engine calculates the engine power and passes the current engine power to `FGPropeller::Calculate(double EnginePower)`.

**Reply by AstroYak** [reply]
Hi Sean, 

Just wanted to say a huge thanks. I managed to iron out all the wrinkles in my system because of you! The breakpoints were the real kicker, but i've got a copy of the textbook and i'm slowly working my way through that too. 

Appreciate your knowledge and swiftness 😊

**Comment by seanmcleod** [comment]
No problem.

**Comment by bcoconni** [comment]
@AstroYak I like very much your flowchart !
Was it handmade or did you use some kind of tooling to generate it ?

**Reply by AstroYak** [reply]
Hey @bcoconni, 

thanks so much! I used this tool to help me draw it but in terms of generation I went through the code line at a time. Probably needs to be verified by somebody much smarter than me!

https://app.diagrams.net/
